pkg/rbd/rbd_journal.go

```diff
@@ -0,0 +1,301 @@
+volume names as requested by the CSI drivers. Hence, these need to be invoked only when the
+respective CSI driver generated snapshot or volume name based locks are held, as otherwise racy
+access to these omaps may end up leaving them in an inconsistent state.
+
+These functions need enough information about cluster and pool (ie, Monitors, Pool, IDs filled in)
+to operate. They further require that the RequestName element of the structure have a valid value
+to operate on and determine if the said RequestName already exists on the backend.
+
+These functions populate the snapshot or the image name, its attributes and the CSI snapshot/volume
+ID for the same when succesful.
+
+These functions also cleanup omap reservations that are stale. I.e when omap entries exist and
+backing images or snapshots are missing, or one of the omaps exist and the next is missing. This is
+because, the order of omap creation and deletion are inverse of each other, and protected by the
+request name lock, and hence any stale omaps are leftovers from incomplete transactions and are
+hence safe to garbage collect.
+*/
+func checkSnapExists(rbdSnap *rbdSnapshot, credentials map[string]string) (found bool, err error) {
+	if err = validateRbdSnap(rbdSnap); err != nil {
+		return false, err
+	}
+
+	key, err := getKey(rbdSnap.AdminID, credentials)
+	if err != nil {
+		return false, err
+	}
+
+	snapUUID, err := snapJournal.CheckReservation(rbdSnap.Monitors, rbdSnap.AdminID, key, rbdSnap.Pool,
+		rbdSnap.RequestName, rbdSnap.RbdImageName)
+	if err != nil {
+		return false, err
+	}
+	if snapUUID == "" {
+		return false, nil
+	}
+	rbdSnap.RbdSnapName = snapJournal.NamingPrefix() + snapUUID
+
+	// Fetch on-disk image attributes
+	err = updateSnapWithImageInfo(rbdSnap, credentials)
+	if err != nil {
+		if _, ok := err.(ErrSnapNotFound); ok {
+			err = snapJournal.UndoReservation(rbdSnap.Monitors, rbdSnap.AdminID, key, rbdSnap.Pool,
+				rbdSnap.RbdSnapName, rbdSnap.RequestName)
+			return false, err
+		}
+		return false, err
+	}
+
+	// found a snapshot already available, process and return its information
+	rbdSnap.SnapID, err = util.GenerateVolID(rbdSnap.Monitors, rbdSnap.AdminID, key, rbdSnap.Pool,
+		rbdSnap.ClusterID, snapUUID, volIDVersion)
+	if err != nil {
+		return false, err
+	}
+
+	klog.V(4).Infof("Found existing snap (%s) with snap name (%s) for request (%s)",
+		rbdSnap.SnapID, rbdSnap.RbdSnapName, rbdSnap.RequestName)
+
+	return true, nil
+}
+
+/*
+Check comment on checkSnapExists, to understand how this function behaves
+
+**NOTE:** These functions manipulate the rados omaps that hold information regarding
+volume names as requested by the CSI drivers. Hence, these need to be invoked only when the
+respective CSI snapshot or volume name based locks are held, as otherwise racy access to these
+omaps may end up leaving the omaps in an inconsistent state.
+*/
+func checkVolExists(rbdVol *rbdVolume, credentials map[string]string) (found bool, err error) {
```

Comment:
found is not used  in this  function

Reply:
Done (removed any further instances of naming return vars as well)